### PR TITLE
Per protocol analysis using pooled log reg based on discretized person-time

### PR DESCRIPTION
**Pull request by alainamstutz**
PR for the per protocol analysis, according to protocol: "We estimate the per-protocol effect of starting metformin monotherapy vs never starting any metformin on the primary outcome by artificially censoring individuals deviating from their assigned treatment strategy (i.e. starting metformin in control group) and using inverse probability of censoring weights to account for this informative censoring. We use the same covariates as for the primary outcome model to estimate the weights, but included time-updated covariate information. We calculate risk ratios with 95% confidence intervals, derived from pooled logistic regression, based on discretized longitudinal time data."

This PR contains the following: 

`PPA_01_add_intervals.R`: 
- 1. Import pre-processed data of a one-person-per-row dataset of eligible participants (data_processed.arrow)
- 2. Restructure the one-person-per-row data into a multiple-event-per-person long format dataset, with time intervals as per study protocol, using the function `fn_expand_intervals.R`

`PPA_02_add_events.R`: 
- 1. Import multiple-event-per-person long format dataset (df_months, which originates from data_processed.arrow)
- 2. Import the separate dataset with stable and dynamic time-updated covariates needed for the per-protocol analysis (data_processed_ppa.arrow -> df_ppa)
- 3. Assign treatment to df_months, using `fn_assign_treatment.R`
- 4. Merge the stable time-updated (first-ever, constant) covariates from df_ppa to df_months. To test edge cases, further adapt dummy data
- 5. Assign the stable time-updated (first-ever, constant) in df_months, using `fn_assign_stable_tu_cov.R`
- 6. Reformatting of the dynamic time-updated covariates in a long-format df_ppa, before merging to df_months. To test edge cases, further adapt dummy data.
- 7. Merge and assign the dynamic time-updated covariates covariates from df_ppa to df_months, using `fn_assign_dynamic_tu_cov.R`
- 8. Fill up the Swiss cheese and reassign the lipid ratio and categorical variables for numeric variables
- 9. Assign and shift outcomes, competing, and censoring events, using `fn_add_and_shift_out_comp_cens_events.R`

`PPA_03_pooled_log_reg.R`: 
- Estimate adjusted but marginal risks per group, risk differences, risk ratios using inverse probability weighting (IPW) and then standardise. Use bootstrapping for 95% CI.
- Create marginal parametric cumulative incidence (risk) curves incl. 95% CI (via bootstrapping)

### CODE REVIEW

**Comment by alainamstutz** [comment]
@pmadleydowd, I noticed the previous PR had become outdated and overly complex to review since it had been open for so long with multiple changes. This PR is much cleaner; and discretizing time in line with the current tutorial rules.

File: analysis/dataset_definition_dm_algo.py
Comment (pmadleydowd) [inline analysis/dataset_definition_dm_algo.py:1]: updated by Alain to help me run on codespaces - fine (and thank you)

File: analysis/dataset_definition_t2dm.py
Comment (pmadleydowd) [inline analysis/dataset_definition_t2dm.py:1]: updated by Alain to help me run on codespaces - fine (and thank you)

File: analysis/functions/fn_assign_stable_tu_cov.R
Comment (pmadleydowd) [inline analysis/functions/fn_assign_stable_tu_cov.R:1]: used in PPA_02_add_events - to review function in context of this

File: analysis/functions/fn_assign_dynamic_tu_cov.R
Comment (pmadleydowd) [inline analysis/functions/fn_assign_dynamic_tu_cov.R:1]: used in PPA_02_add_events - to review function in context of this

File: analysis/PPA_02_add_events.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R:1]: Query: lines 48 to 90 - assigning treatment to df_months
Should you be using the string "cens_date_metfin_start_cont" for date_treat_vars or something else? Its not immediately obvious to me what this is in relation to the treatment definition

File: analysis/functions/fn_assign_treatment.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/functions/fn_assign_treatment.R:1]: looks good other than the above check

File: analysis/PPA_02_add_events.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R:1]: Line 98 and 132 may make it hard to update the covars if we make updates later - you are hard coding these here. 

Suggestion only but you could write a function that defines and stores the names of the stable and dynamic time updating covariates. You would then call these names as vectors - i think this could help but not neccesary to implement if you are fairly certain these wont change at a later date - just be careful with the update if they do

File: analysis/functions/fn_assign_stable_tu_cov.R
Comment (pmadleydowd) [reply analysis/functions/fn_assign_stable_tu_cov.R:1]: rule d - lines 42 to 52 doesnt seem to check whether the covariate event is actually in the interval that the treatment event occurs - i think maybe important to implement so it only applies to the intervals you want it to

File: analysis/PPA_modify_dummy_data_df_months.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_modify_dummy_data_df_months.R:1]: this is a really nice example of exploring edge cases - well done

File: analysis/functions/fn_assign_stable_tu_cov.R
Comment (pmadleydowd) [reply analysis/functions/fn_assign_stable_tu_cov.R:1]: have explored implementation in the data - working as expected

File: project.yaml
Comment (pmadleydowd) [inline project.yaml:1]: addition of PPA actions - ok

File: analysis/PPA_02_add_events.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R:1]: I dont quite understand what treat is doing 
please remind me what exp_bin_treat and cens_bin_metfin_start_cont are

[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_02_add_events.R:1]: - `cens_bin_metfin_start_cont` is the binary flag of `cens_date_metfin_start_cont`, flagging 1 in the month people in control group started metformin (and 0 in all months before and 1 in all months after). However, since `cens_date_metfin_start_cont` is NA in all intervention participants (that's just how the variable comes from ehrQL), all intervention participants receive 0 for their `cens_bin_metfin_start_cont` (see rules of the function)
- `treat` takes the above info and creates a variable whereby all intervention participants are assigned 1 (throughout all months) and 0 in control unless/until they are starting metformin (so, `cens_bin_metfin_start_cont` info for them)
- NOW: I know this is an unnecessary extra step due to how `cens_date_metfin_start_cont` is created. Either I keep this step here, or I adapt `cens_date_metfin_start_cont`. I did not want to adapt the function, because this function will work beautifully if it is fed with a truly time-varying date treatment variable from both arms (i.e. treatment dates from both arms), then it will perfectly get to what I now have as `treat` without extra step.

[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R:1]: many thanks!

[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R:1]: line 207-223: warning message:
In left_join(., df_months, by = "patient_id") :
  Detected an unexpected many-to-many relationship between `x`
and `y`.
ℹ Row 1 of `x` matches multiple rows in `y`.
ℹ Row 1 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set
  `relationship = "many-to-many"` to silence this warning. 
  
  perhaps set relationship = "many-to-many" in left_join to remove errors in log

[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R:1]: similar comment for line 280 - print may be fine - just make sure these checks are clear in the output log

[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R:1]: line 340 - this creates NAs in the filled in dataset. Perhaps the data cleaning (dropping implausible values) should come before filling in the values over time so that the available data can be used

[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R:1]: Note to self - got to line 424 - section 9 - start from here next time

File: analysis/functions/fn_assign_dynamic_tu_cov.R
Comment (alainamstutz) [reply analysis/functions/fn_assign_dynamic_tu_cov.R:1]: You are right to raise this concern. It is not so much about negative/positive  - abs(time_diff_to_treat) ignores the signs -, but it might be that someone has both a covariate date before AND after treatment change date, in same interval. And if the one AFTER treatment change date is closer to the treatment change date, then we have a problem. I think we need to calculate these abs(diff) separately for before/after treatment info change to be on the safe side. Will adapt.

File: analysis/PPA_02_add_events.R
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_02_add_events.R:1]: line 340: You are right - first clean, then assign makes much more sense.

[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_02_add_events.R]: > Comment for section 9 onwards: can any variables be removed from the final datasets to remove storage space. Ok if not possible but worth checking at this stage

Implemented in new section 9 - before running it through the last function.

**Comment by alainamstutz** [comment]
@pmadleydowd : I checked again the PPA_03 script, cleaned up as much as possible. The standardization is now in a function, but the models + predict, I think are better to leave as they are than sourcing out into a function, no?

So, the two remaining things (all the rest please ignore, these were changed due to update from main branch):
1. `fn_fill_forward_incl_baseline.R` to avoid cur_data()
2.  `PPA_03_pooled_log_reg.R`

File: analysis/functions/fn_fill_forward_incl_baseline.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/functions/fn_fill_forward_incl_baseline.R]: from line 435 of dataset_definition_t2dm.py: baseline BMI is the most recent value, within previous 2 years, on or before elig_date_t2dm

[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/functions/fn_fill_forward_incl_baseline.R]: From manuscript  current version 18Nov2025: Rule/step 2: Covariate assignment (reflected in decision tree Option A and illustration Option A)
Definition: If treatment change in interval k, then assign the latest covariate information before treatment change to interval k. Otherwise, carry over the most recent covariate information prior to interval k (any time back until last available value)
Rationale: This ensures that covariate information temporally precedes treatment assignment and in general reflects the latest available information at the start of interval k.

File: analysis/PPA_03_pooled_log_reg.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_03_pooled_log_reg.R]: same as above for lines 379-385

[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_03_pooled_log_reg.R]: Can you explain what is going on in lines 458-477 (Identify target months (indices) for when to extract risk estimates. The final time point is index K, since R indexes from 1 (even though we shifted month to K - 1 above to start at month == 0))

I ask because we get interesting/differing results dependent on the time frame used with the dummy data
image

[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: > Lines 337-347 - i think you need to arrange by patient_id and month before this step to ensure the cumulative product term works as expected

Thanks - that's a really good observation! I think it was working fine because I ordered the dataset further up in the correct way (month within id), but it is unsafe to not ensure it again. I added `arrange(month, .by_group = TRUE) %>%` everywhere where I build cumulative weights along months, inside and outside the bootstrap.

[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: > Can you explain what is going on in lines 458-477 (Identify target months (indices) for when to extract risk estimates. The final time point is index K, since R indexes from 1 (even though we shifted month to K - 1 above to start at month == 0))
> 
> I ask because we get interesting/differing results dependent on the time frame used with the dummy data

Thanks for raising this. As per original protocol, our primary analysis approach (ITT, Cox model) at first defined 1.5 years (550 days) as max. follow-up. We changed this now to 2 years (730 days). So, I need to adapt this. My thinking was I extract two timepoints: (a) the max. follow-up time-point (k=39, a bit over 3 years, that's the max follow-up for the earliest possible inclusion 01.01.2019 to the latest possible follow-up 01.04.2022, hence my dataset is expanded to k) and then (b) the time point that matches the Cox model (primary analysis; see --cut_points=730 in yaml; https://github.com/opensafely/metformin_covid/blob/4e3f8b571f9d0412856156dc715d9d25197c6954/project.yaml#L187). Or is this wrong thinking regarding matching with the Cox?

**Comment by alainamstutz** [comment]
Hi @pmadleydowd , 

I implemented what we discussed last Thursday:

1) Restricted the long format (monthly) dataset to 24 months, our time point of risk estimation. Until now, I only stopped at  study end date (fix for all: 01.04.2022), but I should have stopped expanding at the maximum individual follow-up time (2 years from baseline). This is now implemented:
- PPA_01_add_intervals: I added the max_fup_date to function fn_expand_intervals to incorporate the individual maximum follow-up date (2 years), not only the study end date
- PPA_02_add_events & fn_add_and_shift_out_comp_cens_events: I added the max_fup_date also to function fn_add_and_shift_out_comp_cens_events. 
- PPA_03_pooled_log_reg: K=24, and extract only final time point

2) I added the IPCW for the competing event, to correctly target a controlled direct effect with regards to the competing event (non-covid-death):
- PPA_03_pooled_log_reg: Added IPCW for competing risk and multiplied all 3 weights (treatment, censoring for LTFU, censoring for non-covid-death)

3) I harmonized the coding between stopping for competing event vs stopping for LTFU. Until now, this was inconsistent since that's how the course material from Boston did it. It does not make much difference (since NA will be excluded from estimation), but better to be consistent
- See change in fn_add_and_shift_out_comp_cens_events

4) I updated the Statistical analysis in the protocol: https://docs.google.com/document/d/1Vptfzx-S981m7Th1FIDbAjwe1iUchrXFh0Y6HrVzrDQ/edit?tab=t.ag7eqo6ok9cw This should now match what we do:

"We will also estimate the per-protocol effect of starting metformin monotherapy vs never starting any metformin on the primary outcome, presented as cumulative incidence (risk) curves and 2-year risk ratios and risk differences. We will artificially censor individuals in the control group deviating from their assigned treatment strategy (i.e. starting metformin) and use IPTW to account for this informative censoring. For the  estimation of the treatment weights, we will use the same covariates as for the primary outcome model, but their time-varying versions (where applicable). We account for two additional informative censoring events using inverse probability of censoring weights (IPCW) using the same (time-varying) covariates: The competing risk (non-COVID-deaths) and loss to follow-up (deregistered). As such, this estimates an elimination-of-events estimand, i.e., had no-one died of a non-COVID-related death and had no-one been lost to follow-up. In the competing events framework this relates to a controlled direct effect. For this per-protocol analysis we will apply a parametric approach, using pooled logistic regressions, and discretize the person-time into monthly intervals. We model the follow-up time in these models continuously using linear and quadratic terms (month and month-squared). We will use boostrapping to obtain 95% confidence intervals. For all weights we will calculate the stabilized version, multiply them and truncate them at the 1st and 99th percentile."

5) Why do you think we should add a weighted KM? What additional info does it provide? I wrote: 
"We will estimate the observational analogue of the intention-to-treat effect, i.e., we will ignore the fact that some people may deviate from their strategy (stop metformin in intervention or start metformin in control) and include their outcomes. We will estimate 2-year hazard ratios for each outcome using region-stratified Cox models, with Wald-based 95% confidence intervals, targeting a conditional cause-specific hazard among those still at risk. We will present Cox models adjusted for age and sex only, as well as fully adjusted Cox models, including all of the confounders listed in the covariates list above. Follow-up is censored at outcome, death (competing event), loss to follow-up, administrative end of study or maximum follow-up period (2 years). For the outcomes ‘COVID-related hospitalization or death’ and ‘COVID diagnosis’ non-COVID-related deaths will be treated as a competing event, while for the outcome ‘Long COVID’ all-cause deaths will be treated as a competing event, without censoring weights. We will also estimate absolute risks targeting a marginal (population-average) estimand, including cumulative incidence (risk) curves and 2-year risk ratios and risk differences, with a baseline inverse probability of treatment-weighted (IPTW) Kaplan-Meier approach, without censoring for competing events."
- But not sure the KM adds much, it targets something else, and ideally would also include censoring weights?

6) I read up on the discussion re taking out the interactions in the outcome model `outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)` and understand now that without the interaction this relates to proportional discrete-time hazards (like in standard Cox). But I kept them now in my outcome model, I think it's more realistic to assume a treatment effect change over time. I don't see the value of running the model without interactions, do you? Just to check if it's similar to the Cox model (but then we need to take out the weights, too)?

7) There is 1 change to review left from the previous code review, see unresolved comment further up.

Thanks a million!!

File: analysis/PPA_02_add_events.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_02_add_events.R]: I can't tell whats been updated from the last version - github just says the whole file is new. Is there anything you can do on your end to show the specific changes that have been implemented since last time - this is the same as for the other files

File: analysis/PPA_01_add_intervals.R
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_01_add_intervals.R]: have now viewed this using only changes from last review - update as described - approved

File: analysis/PPA_03_pooled_log_reg.R [hunk 1]
```diff
@@ -556,15 +655,16 @@ plot_cum_risk_treat_ltfu_ci_severecovid <- ggplot(risk_graph,
 
 # Save output -------------------------------------------------------------
 print('Save output')
-# Treatment effect (risk) estimates at max follow-up (39 months)
-write.csv(te_plr_iptw_ipcw_severecovid_39_boot_tbl, file = here::here("output", "te", "pooled_log_reg", "te_plr_iptw_ipcw_severecovid_39_boot_tbl.csv"))
 # Treatment effect (risk) estimates at 24 months (corresponding to Cox model)
-write.csv(te_plr_iptw_ipcw_severecovid_24_boot_tbl, file = here::here("output", "te", "pooled_log_reg", "te_plr_iptw_ipcw_severecovid_24_boot_tbl.csv"))
+write.csv(te_plr_iptw_ipcw_severecovid_boot_tbl, file = here::here("output", "te", "pooled_log_reg", "te_plr_iptw_ipcw_severecovid_boot_tbl.csv"))
```
Comment (pmadleydowd) [inline analysis/PPA_03_pooled_log_reg.R]: just double checking what is being output here (and whether we should consider it equivalent to a Cox model)

Reminder - the exponentiated coefficient for treatment without an interaction effect included in model (i.e. the OR for treatment from the model) will be equivalent to a HR from a Cox model (though we account for time varying censoring through weights which a Cox model cant do) 

The risk ratio using derived risk estimates from the model wont equal the HR

**Comment by pmadleydowd** [comment]
> 4. I updated the Statistical analysis in the protocol: https://docs.google.com/document/d/1Vptfzx-S981m7Th1FIDbAjwe1iUchrXFh0Y6HrVzrDQ/edit?tab=t.ag7eqo6ok9cw This should now match what we do:

Some suggested minor edits in bold and italics below
> "We will also estimate the per-protocol effect of starting metformin monotherapy vs never starting any metformin on the primary outcome, presented as cumulative incidence (risk) curves and 2-year risk ratios and risk differences. We will artificially censor individuals in the control group deviating from their assigned treatment strategy (i.e. starting metformin) and use IPTW to account for this informative censoring. For the estimation of the treatment weights, we will use the same covariates as for the primary outcome model, but _**using**_ their time-varying **_values_**. We account for two additional informative censoring events using inverse probability of censoring weights (IPCW) using the same (time-varying) covariates: The competing risk (non-COVID-deaths) and loss to follow-up (deregistered). As such, this estimates an elimination-of-events estimand, i.e., had no-one died of a non-COVID-related death and had no-one been lost to follow-up. In the competing events framework this relates to a controlled direct effect. For this per-protocol analysis we will apply a parametric approach, using pooled logistic regressions, and discretize the person-time into monthly intervals. We model the follow-up time in these models continuously using linear and quadratic terms (month and month-squared). We will use boostrapping to obtain 95% confidence intervals. For all weights we will calculate the stabilized version, multiply them and truncate them at the 1st and 99th percentile."

Have double checked and we don't include an interaction term between the treatment and time or time squared - this seems to have been removed - is this what we wanted? Its not in the above protocol. 

I think the modelling of follow up time is for the baseline hazard - perhaps need to say this but also double check that I am right

Need to state how weights were stabilised (i.e. using time and lagged treatment)

File: analysis/functions/fn_standardize_risks.R
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/functions/fn_standardize_risks.R]: Thanks, Paul, I did exactly that, looked at my dummy data before and after the change - looked the same.

File: analysis/PPA_03_pooled_log_reg.R [hunk 2]
```diff
@@ -556,15 +655,16 @@ plot_cum_risk_treat_ltfu_ci_severecovid <- ggplot(risk_graph,
```
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: thanks for spotting!

File: analysis/PPA_03_pooled_log_reg.R (hunk 1 above)
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: The outcome model includes an interaction term (line 521). 
I am not after HR, I am after absolute risks, as written in the protocol. 
You think that's an inconsistent PPA approach, with the ITT being a Cox?

**Comment by pmadleydowd** [comment]
> @pmadleydowd , regarding your comment:
> 
> > Have double checked and we don't include an interaction term between the treatment and time or time squared - this seems to have been removed (also comment 6). Note that the risk ratio (from derived risk estimates) wont equal the hazard ratio (just the odds ratio estimate from the pooled logistic regression when we do not include interaction terms with time) as risk and hazard are different quantities - does this make any difference to your decision to include/exclude the interaction term?
> > I think the modelling of follow up time is for the baseline hazard - perhaps need to say this but also double check that I am right
> > Need to state how weights were stabilised (i.e. using time and lagged treatment)
> 
> A few answers:
> 
> 1. We do include an interaction (product term) in the outcome model, see outcome model line 521
> 2. I don't want it to equal the hazard ratio, I am after absolute risks here.
>    => I think this is all fine, right? And corresponds to what is written in the protocol? I think that's another reason for adding weighted KM as an additional ITT approach, an estimand that corresponds better to the PPA approach?
> 3. Good point, let's change that sentence to : "To flexibly model the baseline hazard, we include follow-up time as linear and quadratic terms (month and month-squared) in the outcome model." ok?
> 4. Good point, will add info re stabilised weights, i.e. "...whereby the numerator of the weights is the probability of receiving the treatment that each individual actually received, given past treatment history (i.e. treatment and time variables, but no covariates)"

Meeting 29 Jan 2026: 
1) Alain running PPA with and without interaction term for HR and cumulative incidence separately 
2) discussed estimating both (I think)
3) looks good
4) looks good

File: project.yaml
Comment (pmadleydowd) [inline project.yaml]: updates to PPA_01-PPA_03 - looks sensible

File: analysis/PPA_03_pooled_log_reg.R [hunk 3]
```diff
@@ -56,10 +98,8 @@ print('Define covariates')
 print(covariates_tv_names)
 
 
-# Define bootstraps, primary time point of risk estimation, and time^2 ----------------------------
-print('Define bootstraps, primary time point of risk estimation, and time^2')
```
Comment (pmadleydowd) [inline analysis/PPA_03_pooled_log_reg.R]: removed as dealt with earlier

File: analysis/PPA_03_pooled_log_reg.R [hunk 4]
```diff
@@ -131,10 +171,15 @@ summary(df_months_severecovid$w_treat_stab_trunc)
...
 ## Fit the outcome model -------
 print('Adjust for baseline confounding and time-varying protocol deviation using one time-varying IPTW model: Fit the outcome model')
-treat.severecovid <- parglm(outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr), 
-                                family = binomial(link = 'logit'),
-                                data = df_months_severecovid[df_months_severecovid$censor_LTFU==0 & df_months_severecovid$comp_event == 0,],
-                                weights = df_months_severecovid[df_months_severecovid$censor_LTFU==0 & df_months_severecovid$comp_event == 0,]$w_treat_stab_trunc)
+outcome_formula <- if (include_interaction) {
+  outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)
+} else {
+  outcome ~ exp_bin_treat + month + monthsqr
```
Comment (pmadleydowd) [inline analysis/PPA_03_pooled_log_reg.R]: and as above here

File: analysis/PPA_03_pooled_log_reg.R [hunk 5]
```diff
@@ -517,11 +567,17 @@ te_iptw_ipcw_rd_rr_withCI <- function(data, indices, data_full, covariates_tv_na
...
     # Include the treatment group indicator, the follow-up time (linear and quadratic terms), and product terms between the treatment group indicator and follow-up time
     # Train the model only on individuals who were still at risk of the outcome, i.e. only include individuals who are uncensored and alive
-    # If "I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)" is excluded, then it would mirror the cox model
-    severecovid.treat.ltfu.comp <- parglm(outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr),
-                                    family = binomial(link = 'logit'),
-                                    data = d[d$censor_LTFU==0 & d$comp_event == 0,],
-                                    weights = d[d$censor_LTFU==0 & d$comp_event == 0,]$w_treat_ltfu_comp_stab_trunc)
+    # If "I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)" is excluded, then it mirrors the cox model
+    outcome_formula <- if (include_interaction) {
+      outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)
```
Comment (pmadleydowd) [inline analysis/PPA_03_pooled_log_reg.R]: consider as.formula

**Review by pmadleydowd — CHANGES_REQUESTED** [review]
sorry for delay - some checks needed

also note that the checks say that there is a conflict with main - think this was in the yaml file so should be fairly easy to fix

File: analysis/PPA_03_pooled_log_reg.R [hunk 6]
```diff
@@ -24,6 +24,48 @@ source(here::here("analysis", "functions", "fn_extract_ci_boot.R"))
 source(here::here("analysis", "covariates.R"))
 
 
+# Parse command line arguments --------------------------------------------
+args <- commandArgs(trailingOnly = TRUE)
+
+# Default values
```
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: I think this is line is needed for the if-loop below to capture the args arguments from the command line.

The logic works as follows: 
a) if there is no input yaml args (either for R nor for interaction) => If on real data, R=default 100 and if on dummy data, R=default 5. Output suffix is "". Runs with interaction.
b) if there is only R input (say R=500) in yaml=> If on real data, R=500 and if on dummy data, R=default 5 (hardcoded). Output suffix is "". Runs with interaction.
c) if there is R input (say R=500) and "no_interaction" in yaml => If on real data, R=500 and if on dummy data, R=default 5 (hardcoded). Output suffix is "no_interaction". Runs without interaction.

The second if-loop ensures that if it's run locally (dummy data), then always only R=5, i.e. overwrites R_args. But otherwise there is no overwriting.

File: analysis/PPA_03_pooled_log_reg.R [hunk 7]
```diff
@@ -24,6 +24,48 @@ source(here::here("analysis", "functions", "fn_extract_ci_boot.R"))
...
+# Set R based on environment and arguments
+if (Sys.getenv("OPENSAFELY_BACKEND") %in% c("", "expectations")) {
+  # Local/test environment: always use 5 bootstraps only !!
+  R <- 5
+  if (!is.null(R_arg)) {
+    message("Note: Running on local/test data - ignoring R argument and using R=5 by default")
+  }
+} else {
+  # Real data environment: use argument if provided, otherwise default to 100
+  R <- ifelse(!is.null(R_arg), R_arg, 100)
```
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: Great. I set them to R=200. Or shall we try with say 30 or 50 first to see how long it takes on the real data - before running 200?

File: analysis/PPA_03_pooled_log_reg.R [hunk 8]
```diff
@@ -131,10 +171,15 @@ summary(df_months_severecovid$w_treat_stab_trunc)
 
 ## Fit the outcome model -------
 print('Adjust for baseline confounding and time-varying protocol deviation using one time-varying IPTW model: Fit the outcome model')
-treat.severecovid <- parglm(outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr), 
-                                family = binomial(link = 'logit'),
-                                data = df_months_severecovid[df_months_severecovid$censor_LTFU==0 & df_months_severecovid$comp_event == 0,],
-                                weights = df_months_severecovid[df_months_severecovid$censor_LTFU==0 & df_months_severecovid$comp_event == 0,]$w_treat_stab_trunc)
+outcome_formula <- if (include_interaction) {
+  outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)
```
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: since the formulas are hardcoded and fixed I think it's fine to assign to an R formula object directly

File: analysis/PPA_03_pooled_log_reg.R (hunk 4 above)
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: since the formulas are hardcoded and fixed I think it's fine to assign to an R formula object directly

File: analysis/PPA_03_pooled_log_reg.R (hunk 5 above)
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: Done

File: analysis/PPA_03_pooled_log_reg.R [hunk 9]
```diff
@@ -517,11 +567,17 @@ te_iptw_ipcw_rd_rr_withCI <- function(data, indices, data_full, covariates_tv_na
...
-                                    data = d[d$censor_LTFU==0 & d$comp_event == 0,],
-                                    weights = d[d$censor_LTFU==0 & d$comp_event == 0,]$w_treat_ltfu_comp_stab_trunc)
+    # If "I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)" is excluded, then it mirrors the cox model
+    outcome_formula <- if (include_interaction) {
+      outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)
+    } else {
+      outcome ~ exp_bin_treat + month + monthsqr
+    }
+    
+    severecovid.treat.ltfu.comp <- parglm(outcome_formula,
```
[the remark this answers is not in the document]
Comment (alainamstutz) [reply analysis/PPA_03_pooled_log_reg.R]: See above

File: analysis/PPA_03_pooled_log_reg.R [hunk 10]
```diff
@@ -131,10 +171,15 @@ summary(df_months_severecovid$w_treat_stab_trunc)
...
-                                weights = df_months_severecovid[df_months_severecovid$censor_LTFU==0 & df_months_severecovid$comp_event == 0,]$w_treat_stab_trunc)
+outcome_formula <- if (include_interaction) {
+  outcome ~ exp_bin_treat + month + monthsqr + I(exp_bin_treat*month) + I(exp_bin_treat*monthsqr)
+} else {
+  outcome ~ exp_bin_treat + month + monthsqr
+}
+treat.severecovid <- parglm(outcome_formula,
+                            family = binomial(link = 'logit'),
+                            data = df_months_severecovid[df_months_severecovid$censor_LTFU==0 & df_months_severecovid$comp_event == 0,],
+                            weights = df_months_severecovid[df_months_severecovid$censor_LTFU==0 & df_months_severecovid$comp_event == 0,]$w_treat_stab_trunc)
```
[the remark this answers is not in the document]
Comment (pmadleydowd) [reply analysis/PPA_03_pooled_log_reg.R]: great, thank you for checking!

**Review by pmadleydowd — APPROVED** [review]
